File: packages/typescript/src/features/advanced-capabilities/actions/searchContacts.ts
import { requireActionSpec } from "../../../generated/spec-helpers.ts";
import {
	findKeywordTermMatch,
	getValidationKeywordTerms,
} from "../../../i18n/validation-keywords.ts";
import { logger } from "../../../logger.ts";
import { searchContactsTemplate } from "../../../prompts.ts";
import type { RelationshipsService } from "../../../services/relationships.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";
import { ModelType } from "../../../types/index.ts";
import { composePromptFromState, parseKeyValueXml } from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("SEARCH_CONTACTS");
const SEARCH_KEYWORDS = getValidationKeywordTerms(
	"action.searchContacts.request",
	{
		includeAllLocales: true,
	},
);

interface SearchContactsXmlResult {
	categories?: string;
	searchTerm?: string;
	tags?: string;
	intent?: string;
}

export const searchContactsAction: Action = {
	name: spec.name,
	description: spec.description,
	similes: spec.similes ? [...spec.similes] : [],
	examples: (spec.examples ?? []) as ActionExample[][],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
	): Promise<boolean> => {
		// Check if RelationshipsService is available
		const relationshipsService = runtime.getService(
			"relationships",
		) as RelationshipsService;
		if (!relationshipsService) {
			logger.warn("[SearchContacts] RelationshipsService not available");
			return false;
		}

		// Check if message contains intent to search/list contacts
		const messageText = message.content.text ?? "";
		if (!messageText) return false;
		return findKeywordTermMatch(messageText, SEARCH_KEYWORDS) !== undefined;
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
	): Promise<ActionResult | undefined> => {
		const relationshipsService = runtime.getService(
			"relationships",
		) as RelationshipsService;

		if (!relationshipsService) {
			throw new Error("RelationshipsService not available");
		}

		// Build proper state for prompt composition
		if (!state) {
			state = {
				values: {},
				data: {},
				text: "",
			};
		}

		// Add our values to the state
		state.values = {
			...state.values,
			message: message.content.text,
			senderId: message.entityId,
			senderName: state.values?.senderName || "User",
		};

		// Compose prompt to extract search criteria
		const prompt = composePromptFromState({
			state,
			template: searchContactsTemplate,
		});

		// Use LLM to extract search criteria
		const response = await runtime.useModel(ModelType.TEXT_SMALL, {
			prompt,
			stopSequences: [],
		});

		const parsedResponse = parseKeyValueXml<SearchContactsXmlResult>(response);

		// Build search criteria
		const criteria: {
			categories?: string[];
			tags?: string[];
			searchTerm?: string;
		} = {};

		if (parsedResponse?.categories) {
			criteria.categories = parsedResponse.categories
				.split(",")
				.map((c: string) => c.trim())
				.filter(Boolean);
		}

		if (parsedResponse?.searchTerm) {
			criteria.searchTerm = parsedResponse.searchTerm;
		}

		if (parsedResponse?.tags) {
			criteria.tags = parsedResponse.tags
				.split(",")
				.map((t: string) => t.trim())
				.filter(Boolean);
		}

		// Search contacts
		const contacts = await relationshipsService.searchContacts(criteria);

		// Get entity names for each contact
		const contactDetails = await Promise.all(
			contacts.map(async (contact) => {
				const entity = await runtime.getEntityById(contact.entityId);
				const displayName =
					typeof contact.customFields.displayName === "string"
						? contact.customFields.displayName
						: null;
				return {
					contact,
					entity,
					name: entity?.names[0] || displayName || "Unknown",
				};
			}),
		);

		// Format response
		let responseText = "";

		if (contactDetails.length === 0) {
			responseText = "No contacts found matching your criteria.";
		} else if (parsedResponse?.intent === "count") {
			responseText = `I found ${contactDetails.length} contact${contactDetails.length !== 1 ? "s" : ""} matching your criteria.`;
		} else {
			// Group by category if searching all
			if (!criteria.categories || criteria.categories.length === 0) {
				const grouped: Record<string, typeof contactDetails> = {};
				for (const item of contactDetails) {
					for (const cat of item.contact.categories) {
						const bucket = grouped[cat];
						if (bucket) {
							bucket.push(item);
						} else {
							grouped[cat] = [item];
						}
					}
				}

				const lines: string[] = [];
				lines.push(
					`I found ${contactDetails.length} contact${contactDetails.length !== 1 ? "s" : ""}:`,
					"",
				);

				for (const category in grouped) {
					const items = grouped[category];
					if (!items) continue;
					lines.push(
						`**${category.charAt(0).toUpperCase() + category.slice(1)}s:**`,
					);
					for (const item of items) {
						let line = `- ${item.name}`;
						if (item.contact.tags.length > 0) {
							line += ` [${item.contact.tags.join(", ")}]`;
						}
						lines.push(line);
					}
					lines.push("");
				}
				responseText = lines.join("\n").trim();
			} else {
				const categoryName = criteria.categories[0];
				const lines = [`Your ${categoryName}s:`];
				for (const item of contactDetails) {
					let line = `- ${item.name}`;
					if (item.contact.tags.length > 0) {
						line += ` [${item.contact.tags.join(", ")}]`;
					}
					lines.push(line);
				}
				responseText = lines.join("\n");
			}
		}

		if (callback) {
			await callback({
				text: responseText,
				action: "SEARCH_CONTACTS",
				metadata: {
					count: contactDetails.length,
					criteria,
					success: true,
				},
			});
		}

		return {
			success: true,
			values: {
				count: contactDetails.length,
				criteria,
			},
			data: {
				count: contactDetails.length,
				criteria,
				contacts: contactDetails.map((d) => ({
					id: d.contact.entityId,
					name: d.name,
					categories: d.contact.categories,
					tags: d.contact.tags,
				})),
			},
			text: responseText,
		};
	},
};

File: packages/python/elizaos/features/advanced_capabilities/actions/search_contacts.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import SEARCH_CONTACTS_TEMPLATE
from elizaos.types import (
    Action,
    ActionExample,
    ActionResult,
    Content,
    ModelType,
)
from elizaos.utils.spec_examples import convert_spec_examples
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import (
        HandlerCallback,
        HandlerOptions,
        IAgentRuntime,
        Memory,
        State,
    )

# Get text content from centralized specs
_spec = require_action_spec("SEARCH_CONTACTS")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    return convert_spec_examples(_spec)


@dataclass
class SearchContactsAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, _message: Memory, _state: State | None = None
    ) -> bool:
        relationships_service = runtime.get_service("relationships")
        return relationships_service is not None

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        from elizaos.features.advanced_capabilities.services.relationships import (
            RelationshipsService,
        )

        relationships_service = runtime.get_service("relationships")
        if not relationships_service or not isinstance(relationships_service, RelationshipsService):
            return ActionResult(
                text="Relationships service not available",
                success=False,
                values={"error": True},
                data={"error": "RelationshipsService not available"},
            )

        state = await runtime.compose_state(message, ["RECENT_MESSAGES"])

        prompt = runtime.compose_prompt_from_state(
            state=state,
            template=SEARCH_CONTACTS_TEMPLATE,
        )

        response = await runtime.use_model(ModelType.TEXT_SMALL, {"prompt": prompt})
        parsed = parse_key_value_xml(response)

        categories = None
        tags = None
        search_term = None

        if parsed:
            if parsed.get("categories"):
                categories = [c.strip() for c in str(parsed["categories"]).split(",") if c.strip()]
            if parsed.get("searchTerm"):
                search_term = str(parsed["searchTerm"])
            if parsed.get("tags"):
                tags = [t.strip() for t in str(parsed["tags"]).split(",") if t.strip()]

        contacts = await relationships_service.search_contacts(
            categories=categories,
            tags=tags,
            search_term=search_term,
        )

        contact_details: list[dict[str, str]] = []
        for contact in contacts:
            entity = await runtime.get_entity(str(contact.entity_id))
            name = entity.name if entity and entity.name else "Unknown"
            contact_details.append(
                {
                    "id": str(contact.entity_id),
                    "name": name,
                    "categories": ",".join(contact.categories),
                    "tags": ",".join(contact.tags),
                }
            )

        if not contact_details:
            response_text = "No contacts found matching your criteria."
        else:
            response_text = f"I found {len(contact_details)} contact(s):\n"
            for detail in contact_details:
                response_text += f"- {detail['name']}"
                if detail["categories"]:
                    response_text += f" [{detail['categories']}]"
                response_text += "\n"

        if callback:
            await callback(Content(text=response_text.strip(), actions=["SEARCH_CONTACTS"]))

        return ActionResult(
            text=response_text.strip(),
            success=True,
            values={
                "count": len(contact_details),
            },
            data={
                "count": len(contact_details),
            },
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


search_contacts_action = Action(
    name=SearchContactsAction.name,
    similes=SearchContactsAction().similes,
    description=SearchContactsAction.description,
    validate=SearchContactsAction().validate,
    handler=SearchContactsAction().handler,
    examples=SearchContactsAction().examples,
)
